File: src/bin/mfi_reject_audit.rs
//! Why do accounts our engine flags get rejected? Audit the REAL revert codes.
//!
//! liq_executor logs "chain says healthy at the actionable price" whenever the
//! size-ladder sim returns Some(false) — but simulate_gate maps EVERY custom
//! error to Some(false), not just HealthyAccount(6068). So that one log line
//! actually hides: 6068 (genuinely healthy — if OUR maintenance_health says
//! liquidatable at the SAME on-chain price, that's a health-MATH bug), 6049
//! (Switchboard stale), 6210 (Kamino reserve), size guards, etc.
//!
//! This probe finds every account our maintenance_health flags liquidatable at
//! FRESH on-chain prices (staleness-gated), sims the single-leg liquidate, and
//! tallies the true codes so we can see the real cause distribution.
//!
//! Usage: HELIUS_RPC=<url> [LIQUIDATOR_MA=…] [AUTHORITY=…] cargo run --release --bin mfi_reject_audit

use arb_engine::liquidation::{self as liq, Bank, BankMap, MarginfiAccount, PriceMap};
use arb_engine::marginfi;
use solana_instruction::AccountMeta;
use solana_pubkey::Pubkey;
use std::collections::{HashMap, HashSet};
use std::str::FromStr;
use std::time::Duration;

const MARGINFI_PROGRAM: &str = "MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA";
const MARGINFI_GROUP: &str = "4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8";
const DEFAULT_LIQUIDATOR_MA: &str = "B6e37TbC5n56tWbcgC3RRafUXSuEwRz9ZbhL8Ksro6vD";
const DEFAULT_AUTHORITY: &str = "DYeYAvJSKRokeRkjfgLWKyiT9gwvWPVrT2Sa5xYBFSak";
const USDC_MINT: &str = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
const USDT_MINT: &str = "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB";
const SOL_MINT: &str = "So11111111111111111111111111111111111111112";

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(r) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = r.into_json::<serde_json::Value>() { return Some(v); }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}
fn b64(d: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.decode(d.get(0)?.as_str()?).ok()
}
fn get_multiple(endpoint: &str, keys: &[Pubkey]) -> HashMap<Pubkey, Vec<u8>> {
    let mut out = HashMap::new();
    for chunk in keys.chunks(100) {
        let strs: Vec<String> = chunk.iter().map(|k| k.to_string()).collect();
        let Some(v) = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getMultipleAccounts",
            "params":[strs, {"encoding":"base64"}]})) else { continue };
        for (i, acc) in v["result"]["value"].as_array().into_iter().flatten().enumerate() {
            if let Some(b) = acc.get("data").and_then(b64) { out.insert(chunk[i], b); }
        }
    }
    out
}
fn mint_owner(endpoint: &str, mint: &Pubkey) -> Pubkey {
    rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getAccountInfo",
        "params":[mint.to_string(), {"encoding":"jsonParsed"}]}))
        .and_then(|v| v["result"]["value"]["owner"].as_str().and_then(|s| Pubkey::from_str(s).ok()))
        .unwrap_or_else(|| Pubkey::from_str("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA").unwrap())
}
fn is_debt_mint(m: &Pubkey) -> bool {
    let s = m.to_string();
    s == USDC_MINT || s == USDT_MINT || s == SOL_MINT
}

#[allow(clippy::too_many_arguments)]
fn gate_tx_b64(authority: &Pubkey, liquidator_ma: &Pubkey, tp: &Pubkey, liquidatee: &Pubkey,
    acct: &MarginfiAccount, asset_bank: Pubkey, liab_bank: Pubkey, asset_amount: u64,
    oracle_of: &HashMap<Pubkey, Pubkey>) -> Option<String> {
    use solana_message::{v0, VersionedMessage};
    let mut obs = Vec::new();
    for b in &acct.balances {
        obs.push(AccountMeta::new_readonly(b.bank_pk, false));
        obs.push(AccountMeta::new_readonly(*oracle_of.get(&b.bank_pk)?, false));
    }
    let start = marginfi::start_flashloan(liquidator_ma, authority, 2);
    let liq_ix = marginfi::lending_account_liquidate(&asset_bank, &liab_bank, liquidator_ma, authority,
        liquidatee, tp, asset_amount, oracle_of.get(&asset_bank)?, oracle_of.get(&liab_bank)?, &obs);
    let end_obs = vec![
        AccountMeta::new_readonly(asset_bank, false), AccountMeta::new_readonly(*oracle_of.get(&asset_bank)?, false),
        AccountMeta::new_readonly(liab_bank, false), AccountMeta::new_readonly(*oracle_of.get(&liab_bank)?, false),
    ];
    let end = marginfi::end_flashloan(liquidator_ma, authority, &end_obs);
    let msg = v0::Message::try_compile(authority, &[start, liq_ix, end], &[], solana_hash::Hash::default()).ok()?;
    let tx = solana_transaction::versioned::VersionedTransaction {
        signatures: vec![Default::default()], message: VersionedMessage::V0(msg) };
    use base64::Engine;
    Some(base64::engine::general_purpose::STANDARD.encode(bincode::serialize(&tx).ok()?))
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC").or_else(|_| std::env::var("RPC_HTTP")).expect("HELIUS_RPC");
    let liquidator_ma = Pubkey::from_str(&std::env::var("LIQUIDATOR_MA").unwrap_or_else(|_| DEFAULT_LIQUIDATOR_MA.into())).unwrap();
    let authority = Pubkey::from_str(&std::env::var("AUTHORITY").unwrap_or_else(|_| DEFAULT_AUTHORITY.into())).unwrap();
    let cap: usize = std::env::var("CAP").ok().and_then(|s| s.parse().ok()).unwrap_or(60);

    eprintln!("[audit] scanning marginfi group …");
    let resp = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[MARGINFI_PROGRAM, {"encoding":"base64","dataSlice":{"offset":0,"length":1736},
            "filters":[{"dataSize":liq::MA_SIZE},{"memcmp":{"offset":8,"bytes":MARGINFI_GROUP}}]}]})).expect("scan");
    let accts: Vec<(Pubkey, MarginfiAccount)> = resp["result"].as_array().cloned().unwrap_or_default().iter()
        .filter_map(|e| Some((e["pubkey"].as_str()?.parse().ok()?, MarginfiAccount::decode(&b64(&e["account"]["data"])?)?)))
        .filter(|(_, a): &(Pubkey, MarginfiAccount)| a.balances.iter().any(|b| b.liability_shares > 0.0)).collect();

    let bank_pks: Vec<Pubkey> = accts.iter().flat_map(|(_, a)| a.balances.iter().map(|b| b.bank_pk)).collect::<HashSet<_>>().into_iter().collect();
    let bank_raw = get_multiple(&endpoint, &bank_pks);
    let mut banks: BankMap = HashMap::new();
    let mut oracle_of: HashMap<Pubkey, Pubkey> = HashMap::new();
    for (pk, raw) in &bank_raw { if let Some(bk) = Bank::decode(raw) { oracle_of.insert(*pk, bk.oracle_key); banks.insert(*pk, bk); } }
    let oracle_pks: Vec<Pubkey> = oracle_of.values().copied().collect::<HashSet<_>>().into_iter().collect();
    let slot = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getSlot","params":[{"commitment":"confirmed"}]}))
        .and_then(|v| v["result"].as_u64()).unwrap_or(0);
    let oracle_raw = get_multiple(&endpoint, &oracle_pks);
    // PER-BANK staleness gate (the fix): each bank's on-chain oracle_max_age.
    let mut prices: PriceMap = HashMap::new();
    for (bank_pk, oracle_pk) in &oracle_of {
        let Some(raw) = oracle_raw.get(oracle_pk) else { continue };
        let max_age = banks.get(bank_pk).map(|b| b.oracle_max_age).unwrap_or(0);
        let max_stale = liq::max_stale_slots_for(max_age, liq::DEFAULT_MAX_SB_STALE_SLOTS);
        if let Some(usd) = liq::decode_oracle_price_fresh(raw, slot, max_stale) {
            prices.insert(*bank_pk, usd);
        }
    }

    // Accounts OUR maintenance_health flags liquidatable at FRESH on-chain price,
    // with a wired-debt leg (the ones try_arm would evaluate + reject).
    let mut flagged: Vec<(Pubkey, &MarginfiAccount, Pubkey, Pubkey)> = Vec::new();
    for (pk, a) in &accts {
        let h = liq::maintenance_health(a, &banks, &prices);
        if h.missing > 0 || !h.health.liquidatable() { continue; }
        let asset = a.balances.iter().filter(|b| b.asset_shares > 0.0)
            .max_by(|x, y| {
                let vx = banks.get(&x.bank_pk).and_then(|bk| prices.get(&x.bank_pk).map(|p| x.asset_shares*bk.asset_share_value/10f64.powi(bk.mint_decimals as i32)*p)).unwrap_or(0.0);
                let vy = banks.get(&y.bank_pk).and_then(|bk| prices.get(&y.bank_pk).map(|p| y.asset_shares*bk.asset_share_value/10f64.powi(bk.mint_decimals as i32)*p)).unwrap_or(0.0);
                vx.partial_cmp(&vy).unwrap_or(std::cmp::Ordering::Equal) });
        let debt = a.balances.iter().filter(|b| b.liability_shares > 0.0)
            .find(|b| banks.get(&b.bank_pk).map(|bk| is_debt_mint(&bk.mint)).unwrap_or(false));
        if let (Some(c), Some(d)) = (asset, debt) { flagged.push((*pk, a, c.bank_pk, d.bank_pk)); }
    }
    eprintln!("[audit] {} accounts our maintenance_health flags LIQUIDATABLE at fresh on-chain price (with a wired-debt leg)\n", flagged.len());

    let mut tally: HashMap<String, u32> = HashMap::new();
    let mut examples: HashMap<String, String> = HashMap::new();
    for (pk, a, asset_bank, liab_bank) in flagged.iter().take(cap) {
        let abk = &banks[asset_bank];
        let bal = a.balances.iter().find(|b| b.bank_pk == *asset_bank).unwrap();
        let seize = ((bal.asset_shares * abk.asset_share_value) * 0.02) as u64;
        let tp = mint_owner(&endpoint, &abk.mint);
        let Some(gate) = gate_tx_b64(&authority, &liquidator_ma, &tp, pk, a, *asset_bank, *liab_bank, seize, &oracle_of) else { continue };
        let sim = rpc(&endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"simulateTransaction",
            "params":[gate, {"sigVerify":false,"replaceRecentBlockhash":true,"commitment":"processed","encoding":"base64"}]}));
        let res = sim.as_ref().map(|v| &v["result"]["value"]);
        let err = res.map(|r| &r["err"]);
        let key = match err {
            Some(e) if e.is_null() => "null → FIREABLE (real!)".to_string(),
            Some(e) => {
                let ie = e.get("InstructionError");
                let idx = ie.and_then(|x| x.get(0)).and_then(|x| x.as_u64());
                let code = ie.and_then(|x| x.get(1)).and_then(|c| c.get("Custom")).and_then(|c| c.as_u64());
                match (idx, code) {
                    (Some(1), Some(6068)) => "6068 HealthyAccount  (our math DISAGREES w/ chain at same price → BUG)".to_string(),
                    (Some(1), Some(6049)) => "6049 SwitchboardStalePrice (oracle stale — detection issue)".to_string(),
                    (Some(1), Some(6210)) => "6210 KaminoReserveValidation".to_string(),
                    (Some(1), Some(c)) => format!("in-liquidate Custom({c})"),
                    (Some(i), c) => format!("ix {i} Custom({c:?}) — WIRING?"),
                    _ => format!("other: {e}"),
                }
            }
            None => "rpc-error/no-result".to_string(),
        };
        *tally.entry(key.clone()).or_default() += 1;
        examples.entry(key).or_insert_with(|| pk.to_string());
    }

    println!("\n═══ REJECT-CODE DISTRIBUTION (why flagged accounts don't fire) ═══");
    let mut rows: Vec<_> = tally.into_iter().collect();
    rows.sort_by(|a, b| b.1.cmp(&a.1));
    for (k, n) in &rows {
        println!("  {n:3}  {k}");
        println!("        e.g. {}", examples.get(k).map(|s| s.as_str()).unwrap_or(""));
    }
    println!("\nKEY: 6068 = our health math over-flags vs the chain (a real logic bug to fix).");
    println!("     6049 = stale oracle (detection; the generous 5000-slot gate lets some through).");
    println!("     null = a genuinely fireable account we should have taken.");
}

File: src/bin/mfiRejectAudit.ts
// Port of src/bin/mfi_reject_audit.rs
//
// Why do accounts our engine flags get rejected? Audit the REAL revert codes.
//
// liq_executor logs "chain says healthy at the actionable price" whenever the
// size-ladder sim returns Some(false) — but simulate_gate maps EVERY custom
// error to Some(false), not just HealthyAccount(6068). So that one log line
// actually hides: 6068 (genuinely healthy — if OUR maintenance_health says
// liquidatable at the SAME on-chain price, that's a health-MATH bug), 6049
// (Switchboard stale), 6210 (Kamino reserve), size guards, etc.
//
// This probe finds every account our maintenanceHealth flags liquidatable at
// FRESH on-chain prices (staleness-gated), sims the single-leg liquidate, and
// tallies the true codes so we can see the real cause distribution.
//
// Usage: HELIUS_RPC=<url> [LIQUIDATOR_MA=…] [AUTHORITY=…] npx tsx src/bin/mfiRejectAudit.ts

import 'dotenv/config';
import { PublicKey, TransactionMessage, VersionedTransaction, type AccountMeta } from '@solana/web3.js';
import {
  decodeBank,
  decodeMarginfiAccount,
  decodeOraclePriceFresh,
  maintenanceHealth,
  healthLiquidatable,
  maxStaleSlotsFor,
  DEFAULT_MAX_SB_STALE_SLOTS,
  MA_SIZE,
  type Bank,
  type BankMap,
  type Balance,
  type MarginfiAccount,
  type PriceMap,
} from '../lib/liquidation.js';
import { startFlashloan, endFlashloan, lendingAccountLiquidate } from '../lib/marginfi.js';

const MARGINFI_PROGRAM = 'MFv2hWf31Z9kbCa1snEPYctwafyhdvnV7FZnsebVacA';
const MARGINFI_GROUP = '4qp6Fx6tnZkY5Wropq9wUYgtFxXKwE6viZxFHg3rdAG8';
const DEFAULT_LIQUIDATOR_MA = 'B6e37TbC5n56tWbcgC3RRafUXSuEwRz9ZbhL8Ksro6vD';
const DEFAULT_AUTHORITY = 'DYeYAvJSKRokeRkjfgLWKyiT9gwvWPVrT2Sa5xYBFSak';
const USDC_MINT = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';
const USDT_MINT = 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB';
const SOL_MINT = 'So11111111111111111111111111111111111111112';

async function rpc(endpoint: string, body: unknown): Promise<any | null> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const resp = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await resp.json();
    } catch {
      /* retry */
    }
    await new Promise((r) => setTimeout(r, 400 << attempt));
  }
  return null;
}

function b64(d: any): Buffer | null {
  const s = d?.[0];
  if (typeof s !== 'string') return null;
  try {
    return Buffer.from(s, 'base64');
  } catch {
    return null;
  }
}

async function getMultiple(endpoint: string, keys: PublicKey[]): Promise<Map<string, Buffer>> {
  const out = new Map<string, Buffer>();
  for (let i = 0; i < keys.length; i += 100) {
    const chunk = keys.slice(i, i + 100);
    const strs = chunk.map((k) => k.toBase58());
    const v = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'getMultipleAccounts',
      params: [strs, { encoding: 'base64' }],
    });
    if (v === null) continue;
    const arr: any[] = Array.isArray(v?.result?.value) ? v.result.value : [];
    arr.forEach((acc, idx) => {
      const b = b64(acc?.data);
      if (b) out.set(chunk[idx].toBase58(), b);
    });
  }
  return out;
}

async function mintOwner(endpoint: string, mint: PublicKey): Promise<PublicKey> {
  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getAccountInfo',
    params: [mint.toBase58(), { encoding: 'jsonParsed' }],
  });
  const owner = v?.result?.value?.owner;
  if (typeof owner === 'string') {
    try {
      return new PublicKey(owner);
    } catch {
      /* fall through */
    }
  }
  return new PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
}

function isDebtMint(m: PublicKey): boolean {
  const s = m.toBase58();
  return s === USDC_MINT || s === USDT_MINT || s === SOL_MINT;
}

function gateTxB64(
  authority: PublicKey,
  liquidatorMa: PublicKey,
  tp: PublicKey,
  liquidatee: PublicKey,
  acct: MarginfiAccount,
  assetBank: PublicKey,
  liabBank: PublicKey,
  assetAmount: bigint,
  oracleOf: Map<string, PublicKey>,
): string | null {
  const obs: AccountMeta[] = [];
  for (const b of acct.balances) {
    const oc = oracleOf.get(b.bankPk.toBase58());
    if (!oc) return null;
    obs.push({ pubkey: b.bankPk, isSigner: false, isWritable: false });
    obs.push({ pubkey: oc, isSigner: false, isWritable: false });
  }
  const assetOracle = oracleOf.get(assetBank.toBase58());
  const liabOracle = oracleOf.get(liabBank.toBase58());
  if (!assetOracle || !liabOracle) return null;
  const start = startFlashloan(liquidatorMa, authority, 2n);
  const liqIx = lendingAccountLiquidate(assetBank, liabBank, liquidatorMa, authority, liquidatee, tp, assetAmount, assetOracle, liabOracle, obs);
  const endObs: AccountMeta[] = [
    { pubkey: assetBank, isSigner: false, isWritable: false },
    { pubkey: assetOracle, isSigner: false, isWritable: false },
    { pubkey: liabBank, isSigner: false, isWritable: false },
    { pubkey: liabOracle, isSigner: false, isWritable: false },
  ];
  const end = endFlashloan(liquidatorMa, authority, endObs);
  const zeroBlockhash = new PublicKey(Buffer.alloc(32)).toBase58();
  const msg = new TransactionMessage({
    payerKey: authority,
    recentBlockhash: zeroBlockhash,
    instructions: [start, liqIx, end],
  }).compileToV0Message([]);
  const tx = new VersionedTransaction(msg);
  tx.signatures = [new Uint8Array(64)];
  return Buffer.from(tx.serialize()).toString('base64');
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (!endpoint) throw new Error('HELIUS_RPC');
  const liquidatorMa = new PublicKey(process.env.LIQUIDATOR_MA ?? DEFAULT_LIQUIDATOR_MA);
  const authority = new PublicKey(process.env.AUTHORITY ?? DEFAULT_AUTHORITY);
  const cap = Number.parseInt(process.env.CAP ?? '', 10) || 60;

  console.error('[audit] scanning marginfi group …');
  const resp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [
      MARGINFI_PROGRAM,
      {
        encoding: 'base64',
        dataSlice: { offset: 0, length: 1736 },
        filters: [{ dataSize: MA_SIZE }, { memcmp: { offset: 8, bytes: MARGINFI_GROUP } }],
      },
    ],
  });
  if (resp === null) throw new Error('scan');
  const entries: any[] = Array.isArray(resp?.result) ? resp.result : [];
  const accts: Array<[PublicKey, MarginfiAccount]> = [];
  for (const e of entries) {
    const pkStr = e?.pubkey;
    if (typeof pkStr !== 'string') continue;
    const raw = b64(e?.account?.data);
    if (!raw) continue;
    let pk: PublicKey;
    try {
      pk = new PublicKey(pkStr);
    } catch {
      continue;
    }
    const a = decodeMarginfiAccount(raw);
    if (!a) continue;
    if (!a.balances.some((b) => b.liabilityShares > 0.0)) continue;
    accts.push([pk, a]);
  }

  const bankPkSet = new Map<string, PublicKey>();
  for (const [, a] of accts) for (const b of a.balances) bankPkSet.set(b.bankPk.toBase58(), b.bankPk);
  const bankPks = [...bankPkSet.values()];
  const bankRaw = await getMultiple(endpoint, bankPks);
  const banks: BankMap = new Map();
  const oracleOf = new Map<string, PublicKey>();
  for (const [pkStr, raw] of bankRaw) {
    const bk = decodeBank(raw);
    if (bk) {
      oracleOf.set(pkStr, bk.oracleKey);
      banks.set(pkStr, bk);
    }
  }
  const oraclePkSet = new Map<string, PublicKey>();
  for (const oc of oracleOf.values()) oraclePkSet.set(oc.toBase58(), oc);
  const oraclePks = [...oraclePkSet.values()];
  const slotResp = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getSlot',
    params: [{ commitment: 'confirmed' }],
  });
  const slot = BigInt(typeof slotResp?.result === 'number' ? slotResp.result : 0);
  const oracleRaw = await getMultiple(endpoint, oraclePks);
  // PER-BANK staleness gate (the fix): each bank's on-chain oracle_max_age.
  const prices: PriceMap = new Map();
  for (const [bankPk, oraclePk] of oracleOf) {
    const raw = oracleRaw.get(oraclePk.toBase58());
    if (!raw) continue;
    const maxAge = banks.get(bankPk)?.oracleMaxAge ?? 0;
    const maxStale = maxStaleSlotsFor(maxAge, DEFAULT_MAX_SB_STALE_SLOTS);
    const usd = decodeOraclePriceFresh(raw, slot, maxStale);
    if (usd !== null) prices.set(bankPk, usd);
  }

  // Accounts OUR maintenance_health flags liquidatable at FRESH on-chain price,
  // with a wired-debt leg (the ones try_arm would evaluate + reject).
  const flagged: Array<{ pk: PublicKey; a: MarginfiAccount; assetBank: PublicKey; liabBank: PublicKey }> = [];
  for (const [pk, a] of accts) {
    const h = maintenanceHealth(a, banks, prices);
    if (h.missing > 0 || !healthLiquidatable(h.health)) continue;
    const assetValue = (b: Balance): number => {
      const bk = banks.get(b.bankPk.toBase58());
      const px = prices.get(b.bankPk.toBase58());
      if (!bk || px === undefined) return 0.0;
      return (b.assetShares * bk.assetShareValue) / 10 ** bk.mintDecimals * px;
    };
    let asset: Balance | undefined;
    for (const b of a.balances.filter((x) => x.assetShares > 0.0)) {
      if (asset === undefined || assetValue(b) > assetValue(asset)) asset = b;
    }
    const debt = a.balances.filter((b) => b.liabilityShares > 0.0).find((b) => {
      const bk = banks.get(b.bankPk.toBase58());
      return bk ? isDebtMint(bk.mint) : false;
    });
    if (asset !== undefined && debt !== undefined) {
      flagged.push({ pk, a, assetBank: asset.bankPk, liabBank: debt.bankPk });
    }
  }
  console.error(
    `[audit] ${flagged.length} accounts our maintenance_health flags LIQUIDATABLE at fresh on-chain price (with a wired-debt leg)\n`,
  );

  const tally = new Map<string, number>();
  const examples = new Map<string, string>();
  for (const { pk, a, assetBank, liabBank } of flagged.slice(0, cap)) {
    const abk = banks.get(assetBank.toBase58())!;
    const bal = a.balances.find((b) => b.bankPk.equals(assetBank))!;
    const seize = BigInt(Math.trunc(bal.assetShares * abk.assetShareValue * 0.02));
    const tp = await mintOwner(endpoint, abk.mint);
    const gate = gateTxB64(authority, liquidatorMa, tp, pk, a, assetBank, liabBank, seize, oracleOf);
    if (gate === null) continue;
    const sim = await rpc(endpoint, {
      jsonrpc: '2.0',
      id: 1,
      method: 'simulateTransaction',
      params: [gate, { sigVerify: false, replaceRecentBlockhash: true, commitment: 'processed', encoding: 'base64' }],
    });
    const res = sim?.result?.value;
    const err = res?.err;
    let key: string;
    if (err === null) {
      key = 'null → FIREABLE (real!)';
    } else if (err !== undefined) {
      const ie = err?.InstructionError;
      const idx: number | undefined = Array.isArray(ie) ? ie[0] : undefined;
      const code: number | undefined = Array.isArray(ie) ? ie[1]?.Custom : undefined;
      if (idx === 1 && code === 6068) {
        key = '6068 HealthyAccount  (our math DISAGREES w/ chain at same price → BUG)';
      } else if (idx === 1 && code === 6049) {
        key = '6049 SwitchboardStalePrice (oracle stale — detection issue)';
      } else if (idx === 1 && code === 6210) {
        key = '6210 KaminoReserveValidation';
      } else if (idx === 1 && code !== undefined) {
        key = `in-liquidate Custom(${code})`;
      } else if (idx !== undefined) {
        key = `ix ${idx} Custom(${code ?? 'null'}) — WIRING?`;
      } else {
        key = `other: ${JSON.stringify(err)}`;
      }
    } else {
      key = 'rpc-error/no-result';
    }
    tally.set(key, (tally.get(key) ?? 0) + 1);
    if (!examples.has(key)) examples.set(key, pk.toBase58());
  }

  console.log('\n═══ REJECT-CODE DISTRIBUTION (why flagged accounts don\'t fire) ═══');
  const rows = [...tally.entries()].sort((a, b) => b[1] - a[1]);
  for (const [k, n] of rows) {
    console.log(`  ${String(n).padStart(3)}  ${k}`);
    console.log(`        e.g. ${examples.get(k) ?? ''}`);
  }
  console.log('\nKEY: 6068 = our health math over-flags vs the chain (a real logic bug to fix).');
  console.log('     6049 = stale oracle (detection; the generous 5000-slot gate lets some through).');
  console.log('     null = a genuinely fireable account we should have taken.');
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
